rfcs/038-unify-shared-dependencies.md

```diff
@@ -0,0 +1,112 @@
+2. Identify libraries that are shared across all (or majority - see `Drawbacks`) of subprojects
+3. Create `requiremnts-common.txt` file in root folder of amundsen repo containing pinned dependencies for shared libraries
+4. Remove libraries listed in `requirements-common.txt` from `requirements.txt` of each subdirectory
+5. Create `requirements-dev.txt` file in root folder of amundsen repo containing pinned dependencies for shared dev libraries (mypy, flake8, etc.)
+6. Remove libraries listed in `requirements-dev.txt` from `requirements.txt` of each subdirectory
+7. Create symbolic link to `requirements-common.txt` and `requirements-dev.txt` in each subdirectory
+8. Shift context of building docker images to root folder (otherwise we cannot share `requirements-common.txt` and `requirements-dev.txt` symlinks across subdirectories)
+9. Extend every `setup.py` to contain following entries (`extras_require` might contain more keys)"
+```python
+install_require=requirements + requirements_common
+extras_require(dict(all=requirements + requirements_common + requirements_dev, dev=requirements_dev)),
+``` 
+Where:
+    - `requirements` - requirements.txt specific for subdirectory
+    - `requirements_common` - requirements-common.txt /symlink/ from project root
+    - `requirements_dev` - requirements-dev.txt /symlink/ from project root)
+10. Introduce `install_deps` action in Makefile to be used by users/in cicd, doing:
+```python
+pip3 install ".[all]"
+```
+10. In CI/CD apart from `make install` do also `pip3 uninstall amundsen-common && pip3 install ../common` (apart from common subdirectory which would not re-install ../common)
+
+### Testing
+We test this with `make test` on each subproject + building Amundsen locally and testing functionalities manually relying on test data
+
+## How We Communicate This
+
+- We'll promote this RFC frequently on Amundsen Slack and other social media channels, to get feedback from the community.
+- During the community meetings, update on the progress
+- Update docs for each component
+
+## Drawbacks
+
+- As we don't have 100% test coverage it's difficult to asses complete impact of evening out dependencies so we might enter period of ironing out potential issues
+- Using symlinks might be tricky for users using Windows
+- There are dependencies that are shared by some but not all libraries. For those we have two choices:
+    - Include them in `requirements-common.txt` (preferred by the author) - this will make some packages contain deps that are not used
+    - Populate `requirements-common.txt` only with deps used by all packages - this will still leave a lot of room for inconsistencies
+- It might be difficult to keep the dependencies balanced especially if some custom entry in one of subdirs `requirements.txt` requires different version of lib from `requirements-common.txt`
+
+## Alternatives
+
+- Using pip functionality `constraint` but I've identified this as less intuitive for end users as:
```

Comment:
I'm not familiar enough with python packaging to have an answer here, but it is a bit of a shame we can't use constraint files -- it feels like the "correct" solution. But I do agree it's not intuitive, and even if we published build scripts everywhere, many people wouldn't use them because there is an expectation you can just `pip install -r requirements` and be off to the races

If instead of including requirements.txt, we only used `setup.py` deps + constraint file, i wonder if it's possible to automatically bring in the constraint file in setup.py or setup.cfg ?

Reply:
as far as I know (and from research I did) there is no way to make setup.py constraints aware. what you would do in such case is just list shared dependencies without pinning versions and then, when doing pip install you supply pinned versions using `--constraint`  parameter. what's important is that --constraint doesn't force installation of additional deps - if something is listed in constraints file but not in install_requires then it won't be installed

Reply:
to be completely precise - I also was thinking about making clearing `setup.py` files out of deps and pull them by reading `requirements*txt` files